tutorials/optimize/opt_gemm.py

```diff
@@ -136,30 +141,115 @@
+func = tvm.build(s, [A, B, C], name = 'mmult')
+assert func
+
 c = tvm.nd.array(numpy.zeros((N, N), dtype = dtype), tvm.cpu(0))
+func(a, b, c)
+numpy.testing.assert_allclose(c.asnumpy(), answer, rtol=1e-5)
+
+evaluator = func.time_evaluator(func.entry_name, tvm.cpu(0), number=5)
 print('Opt3: %f' % evaluator(a, b, c).mean)
 
+################################################################################################
+# Here is the generated IR after vectorization.
+
+print(tvm.lower(s, [A, B, C], simple_mode=True))
+
+###################################################################################################
+# Loop Permutation
+# -------------
+# If we look at the above IR, we can see the inner loop row data is vectorized and
+# B is transformed into PackedB. The traversal of PackedB is sequential now.
+# So we will look at the access pattern of A. In current schedule, A is accessed column by column
+# which is not cache friendly. If we change the nested loop order of k and inner row index,
+# the access pattern for A matrix is more cache friendly.
+
+s = tvm.create_schedule(C.op)
+xo, yo, xi, yi = s[C].tile(C.op.axis[0], C.op.axis[1], bn, bn)
+s[C].reorder(xo, yo, xi, k, yi)
+
+# Vectorization
+s[C].vectorize(yi)
+
+func = tvm.build(s, [A, B, C], name = 'mmult')
+assert func
+
+c = tvm.nd.array(numpy.zeros((N, N), dtype = dtype), tvm.cpu(0))
+func(a, b, c)
+numpy.testing.assert_allclose(c.asnumpy(), answer, rtol=1e-5)
+
+evaluator = func.time_evaluator(func.entry_name, tvm.cpu(0), number=5)
+print('Opt4: %f' % evaluator(a, b, c).mean)
+
+################################################################################################
+# Here is the generated IR after loop permutation.
+
+print(tvm.lower(s, [A, B, C], simple_mode=True))
+
+###################################################################################################
+# Parallel
+# -------------
+# Futhermore, we can also utilize multi-core processors to parallelize computation.
+
+s = tvm.create_schedule(C.op)
+xo, yo, xi, yi = s[C].tile(C.op.axis[0], C.op.axis[1], bn, bn)
+s[C].reorder(xo, yo, xi, k, yi)
+s[C].vectorize(yi)
+
+# parallel
+s[C].parallel(xo)
+
+func = tvm.build(s, [A, B, C], name = 'mmult')
+assert func
+
+c = tvm.nd.array(numpy.zeros((N, N), dtype = dtype), tvm.cpu(0))
+func(a, b, c)
+numpy.testing.assert_allclose(c.asnumpy(), answer, rtol=1e-5)
+
+evaluator = func.time_evaluator(func.entry_name, tvm.cpu(0), number = 5)
+opt5_time = evaluator(a, b, c).mean
+print('Opt5: %f' % opt5_time)
+
 ##################################################################################################
 # Summary
 # -------
-# After applying three main tricks, we can achieve almost 90% performance of numpy.
-# Further observation is required to catch up with the performance of numpy.
+# After applying the above simple optimizations with only 9 lines of code,
+# our generated code can achieve 30% of numpy performance with Apple implemented BLAS.
```

Comment:
Finally, please refer to a similar implementation from @javelinjs for more insight. https://github.com/javelinjs/topi-intel/blob/master/opt_gemm.py

Reply:
I have no idea why cache_write for C actually hurts the performance in my schedule, if anyone can give some hints.
BTW, I just changed my github username so as to be more easily mentioned.

Reply:
yzhliu's code is very insightfuly, I probably will have time later this week to investigate more.